### PR TITLE
Publish GitHub Release on Tag

### PR DESCRIPTION
**Pull request by indyfree**
Towards https://github.com/giantswarm/giantswarm/issues/4848

Adds CI for publishing a GH release using `ghr`(https://github.com/tcnksm/ghr) when pushing a tag. Also lints the shell script.

### CODE REVIEW

**Review by puja108 — APPROVED** [review]

**Review by xh3b4sd — COMMENTED** [review]
I would like to have the knowledge share and agreement with @giantswarm/sig-releng before merging this. All of this looks new and different. I cannot judge right now and would like to have more reasoning and opinions from others.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,5 +1,15 @@
 version: 2
 jobs:
+  lint:
+    docker:
+      - image: koalaman/shellcheck-alpine
```
[the remark this answers is not in the document]
Comment (indyfree) [reply .circleci/config.yml]: Oh yes totally, good catch!

**Comment by indyfree** [comment]
Also @marians implemented a similar approach for gsctl in this (and the following) PRs:
https://github.com/giantswarm/gsctl/pull/316

File: .circleci/config.yml [hunk 2]
```diff
@@ -18,8 +28,36 @@ jobs:
...
+
+      - run:
+          name: "Package Helm Chart"
+          command: |
+            ./ci-scripts/package.sh ${CIRCLE_PROJECT_REPONAME} ${CIRCLE_TAG}
+
+      - run:
+          name: "Publish Release on GitHub"
+          command: |
+            ghr -t ${PERSONAL_ACCESS_TOKEN} -u giantswarm -r ${CIRCLE_PROJECT_REPONAME} -c ${CIRCLE_SHA1} ${CIRCLE_TAG} "${CIRCLE_PROJECT_REPONAME}-chart-${CIRCLE_TAG}.tgz"
```
Comment (indyfree) [inline .circleci/config.yml]: @marians `ghr` might be a very useful tool for `gsctl` releases too!

File: .circleci/config.yml [hunk 3]
```diff
@@ -18,8 +28,36 @@ jobs:
...
 workflows:
   version: 2
   build-deploy:
     jobs:
-      - build
+      - lint
+      - build:
+          filters:
+            tags:
+              only: /^\d+\.\d+\.\d+$/
```
Comment (kopiczko) [inline .circleci/config.yml]: Let's pease use tag in format `v1.2.3`. This is not only kind of standard, but also works well with go modules.

```suggestion
              only: /v^\d+\.\d+\.\d+$/
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -18,8 +28,36 @@ jobs:
...
+      - publish-github-release:
+          requires:
+            - build
+          filters:
+            branches:
+              ignore: /.*/
+            tags:
+              only: /^\d+\.\d+\.\d+$/
```
Comment (kopiczko) [inline .circleci/config.yml]: Same here.

```suggestion
              only: /v^\d+\.\d+\.\d+$/
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -18,8 +28,36 @@ jobs:
     - store_test_results:
         path: /tmp/results
 
+  publish-github-release:
+    docker:
+      - image: cibuilds/github:0.12
```
Comment (kopiczko) [inline .circleci/config.yml]: Should be also retagged.

File: ci-scripts/package.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+readonly PROJECT=$1
+readonly VERSION=$2
```
Comment (kopiczko) [inline ci-scripts/package.sh]: TIL `readonly`

**Comment by puja108** [comment]
Ok with @giantswarm/sig-releng review/sync/approval on this, and we are currently not in a rush on this. If needed I can also join and explain the why and how of this in connection to the general release process we envision with app operator. I also mentioned in the PR for the VOO story that this process might be related and the general successor of the current release process at least for full releases (not feature-branch foo that is still kept on quay). We need a bit of a 3-way sync between batman, magic, and releng here as we are all working on the future of releng with these tools and I do want to avoid 3 different solutions to the same problem

File: ci-scripts/package.sh [hunk 7]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
```
Comment (kopiczko) [inline ci-scripts/package.sh]: What are plans for this file. Do we want to have that in all repos? Do we want to have that built into architect? Or maybe we want to make it a cricleci orb.

File: .circleci/config.yml (hunk 3 above)
Comment (indyfree) [reply .circleci/config.yml]: @puja108 and @giantswarm/sig-releng do we also want to have our Chart versions prefixed with `v` then, or only the tags?

File: ci-scripts/package.sh (hunk 7 above)
Comment (puja108) [reply ci-scripts/package.sh]: yes, we need to sync with VOO and sig releng on this matter in general.

+1 on having it somewhere central.

We also talked about checking out GH actions, and see if we really need the external dependency on circle CI (even as orb) for this simple functionality or if it living within GH would make more sense.

File: .circleci/config.yml (hunk 3 above)
Comment (puja108) [reply .circleci/config.yml]: hmm, good call, now that you say it upstream standard is not having a v there, so might be nicer to not have it, just cause we do that internally a lot and it's easier with modules doesn't mean it needs to be in the chart and with that the URL of the tgz

Comment (rossf7) [reply .circleci/config.yml]: For the `v` prefix I'd prefer it if we were consistent. Having to remember to use it sometimes and not others is annoying and error prone.

File: ci-scripts/package.sh (hunk 7 above)
Comment (rossf7) [reply ci-scripts/package.sh]: Investigating GH actions and Orbs makes sense to me. As then maybe we don't need the file. I agree we need to sync with VOO and sig-releng here.

File: .circleci/config.yml (hunk 3 above)
Comment (kopiczko) [reply .circleci/config.yml]: It doesn't need to be in the chart, but I insist to have git tags in `v1.2.3` format. Otherwise we'll have to make tooling aware of both.

Comment (kopiczko) [reply .circleci/config.yml]: I'd argue that if both things are following standards it's even easier to remember.

**Review by kopiczko — CHANGES_REQUESTED** [review]
Blocking till we have tag format sorted. This is important for me.

File: .circleci/config.yml (hunk 3 above)
Comment (puja108) [reply .circleci/config.yml]: you're right, when I look at most tags/releases on github in kubernetes world they do carry `v`, and indeed with chart versions internally that is not the case, so that is the only expection here

TLDR standards are v everywhere but in the version field of a yaml (kinda :wink: )

File: ci-scripts/package.sh (hunk 7 above)
Comment (puja108) [reply ci-scripts/package.sh]: and it is ok to have it first in a script/file and then move that around to a better implementation (also with more experience on where the process does not work for us in real life then)

File: .circleci/config.yml (hunk 3 above)
Comment (TheoBrigitte) [reply .circleci/config.yml]: I think the regex here should be `/^v\d+\.\d+\.\d+$/` (`v` after `^`)

Comment (kopiczko) [reply .circleci/config.yml]: Thanks @TheoBrigitte. Updated.

Comment (puja108) [reply .circleci/config.yml]: Yeah, let's go with that for now. Everyone ok with that?

**Review by rossf7 — APPROVED** [review]
LGTM

File: .circleci/config.yml (hunk 3 above)
Comment (indyfree) [reply .circleci/config.yml]: Cool. Sorted and updated. PTAL @kopiczko

**Review by kopiczko — APPROVED** [review]
LGTM

File: .circleci/config.yml [hunk 8]
```diff
@@ -18,8 +28,36 @@ jobs:
...
 workflows:
   version: 2
   build-deploy:
     jobs:
-      - build
+      - lint
+      - build:
+          filters:
+            tags:
+              only: /^v\d+\.\d+\.\d+$/
```
[the remark this answers is not in the document]
Comment (indyfree) [reply .circleci/config.yml]: Its not build only on tags, but **also** on tags with that tag format. Circle uses whitelisting for tags, in contrary to branches. So the syntax looks just a bit weird.

[the remark this answers is not in the document]
Comment (kopiczko) [reply .circleci/config.yml]: Thanks for clarifying. I got confused.
